In [1]:
from __future__ import division , print_function
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.lines as mlines
import pandas as pd
#import seaborn as sns
import gmaps
import gmaps.datasets
import os
%matplotlib inline

ModuleNotFoundError: No module named 'gmaps'

In [2]:
# Se lee los datos
data = pd.read_csv("B63-2011-04-03_2011-05-03.csv", parse_dates=['timestamp'])

In [3]:
# API Key de google para mapas
gmaps.configure(api_key="AIzaSyCgDcCbAlDWczyxOFcLI3Zd4IboRl3p3jQ")

# Vizualización de la ruta B63

In [4]:
locations = data[["latitude", "longitude"]][:100000]
m = gmaps.Map()
m.add_layer(gmaps.heatmap_layer(locations, weights=None))
m

In [3]:
def gen_dates(dataset_timestamp):
    "Genera las fechas del mes, retorna una lista de strings numpy"
    dates = dataset_timestamp.values
    dates_list = []
    for i,date in enumerate(dates):    
        dates_list.append(str(date)[:10])
    dates_strings = np.unique(np.array(dates_list))
    return dates_strings



def info_per_day(days, vehicle_id, dataset):
    "Retorna un diccionario con key los ids de los buses y cada uno tendrá una lista como value donde esta \
    la intensidad horaria de cada día durante el mes"
   
    info_bus_month = {}
    intensity = []
    
    for i in vehicle_id:
        for day in days:
            next_day = str(pd.datetime(int(day[:4]),int(day[5:7]),int(day[8:11]))  + pd.Timedelta('1 days'))
            route_day = dataset[(dataset['timestamp'] >= str(day)) & (dataset['timestamp'] < next_day)]
            route_day_bus = route_day[route_day.vehicle_id==i]
            
            if ((len(route_day_bus.values)) == 0) == True:
                intensity.append(0)
                
            else:
                start = route_day_bus.timestamp.min()
                end = route_day_bus.timestamp.max()
                working_hours = end - start
                intensity.append(float(working_hours.seconds/3600))
                
        info_bus_month[i] = intensity
        intensity = []
    return info_bus_month
    

# Cantidad y id's de los buses

In [4]:
bus_ids, counts = np.unique(data.vehicle_id, return_counts=True)

In [9]:
print ("Número de buses: ",len(data.vehicle_id.unique()))

Número de buses:  30


In [10]:
for i,bus in enumerate(bus_ids):
    print ("Bus ",bus_ids[i],"Tiene",counts[i],"registros")

Bus  7560 Tiene 30343 registros
Bus  7561 Tiene 5737 registros
Bus  7562 Tiene 29463 registros
Bus  7563 Tiene 25533 registros
Bus  7564 Tiene 32787 registros
Bus  7565 Tiene 32936 registros
Bus  7566 Tiene 12825 registros
Bus  7568 Tiene 29049 registros
Bus  7569 Tiene 34565 registros
Bus  7570 Tiene 40224 registros
Bus  7571 Tiene 37196 registros
Bus  7572 Tiene 28882 registros
Bus  7573 Tiene 42952 registros
Bus  7574 Tiene 38333 registros
Bus  7575 Tiene 31752 registros
Bus  7576 Tiene 23647 registros
Bus  7577 Tiene 24529 registros
Bus  7578 Tiene 33365 registros
Bus  7579 Tiene 21779 registros
Bus  7580 Tiene 37243 registros
Bus  7581 Tiene 22497 registros
Bus  7582 Tiene 30875 registros
Bus  7583 Tiene 29792 registros
Bus  7584 Tiene 19804 registros
Bus  7585 Tiene 29412 registros
Bus  7586 Tiene 22030 registros
Bus  7587 Tiene 24798 registros
Bus  7588 Tiene 32253 registros
Bus  7589 Tiene 33772 registros
Bus  9864 Tiene 2567 registros


<h4>Intensidad horaria</h4>

In [5]:
hourly_intensity = info_per_day(gen_dates(data.timestamp),bus_ids,data)

In [12]:
xe=pd.DataFrame.from_dict(hourly_intensity)
xsum = xe.cumsum()
plt.figure();
ax = xsum.plot(figsize=(20, 10))
ax.set_ylabel('Horas totales')
ax.set_xlabel('Dias del mes')
ax.set_title('Suma total de horas en el mes (Intensidad horaria)')

<h4>Intensidad por día sin suma</h4>

In [13]:
xe.plot(subplots=True,  figsize=(20, 50),sharex=False);

<p>Se puede concluir que algunos buses son fijos, y otro solo entran en funcionamiento en horas pico, como apoyo
también podría decirse que en ese mes algunos automotores tuvieron poblemas mecánicos por ejemplo el bus con
el código 9864 tuvo muy poca intensidad a lo largo del mes, quizás averiado</p>

In [6]:
3

3

<h4>Nota:</h4>
<p>Encontre que en las tomas de datos se hacen 30 segundos, por ejemplo cuando el bus esta en la parada para empezar
la ruta del día se alcanzan a tomar varios timestamp antes de que inicie con la ruta, osea que se pueden limpiar los 
datos eliminando las las mismas filas tocaría ver si las medidas de tiempo se hacen también en el recorrido o
solo en las paradas</p>

<h4>Distancia recorrida promedio de la ruta de ida y vuelta (en metros)</h4>

In [7]:
def treatment_of_dataset(days, vehicle_id, dataset,data_0,data_1):
    data_0 = data_0.copy()
    data_1 = data_1.copy()
    column_distance_mean_0, column_distance_mean_1, column_distance_mean = [],[],[]
    for i in vehicle_id:
        for day in days:
            next_day = str(pd.datetime(int(day[:4]),int(day[5:7]),int(day[8:11]))  + pd.Timedelta('1 days'))
            route_day = dataset[(dataset['timestamp'] >= str(day)) & (dataset['timestamp'] < next_day)]
            route_day_bus = route_day[route_day.vehicle_id==i]
            direction_data_1 = route_day_bus[route_day_bus.direction_id == 1]
            column_1 = np.ones(direction_data_1.shape[0])
            mean_1 = direction_data_1[direction_data_1.stop_sequence == 57][['shape_dist_traveled']].mean()
            column_1 = column_1 * float(mean_1)
            direction_data_0 = route_day_bus[route_day_bus.direction_id == 0]
            column_0 = np.ones(direction_data_0.shape[0])
            mean_0 = direction_data_0[direction_data_0.stop_sequence == 56][['shape_dist_traveled']].mean()
            column_0 = column_0 * float(mean_0)
            
            column_distance_mean_0 += list(column_0)
            column_distance_mean_1 += list(column_1)
        
    data_0['average_distance'] = np.array( column_distance_mean_0)
    data_1['average_distance'] = np.array( column_distance_mean_1)
    
    # Agregando la columna de distancia promedio de viaje según el sentido
    result = [data_0,data_1]
    datasetr = pd.concat(result,ignore_index=True)
    
    #Eliminando las dos columnas, creemos que aportan poco a la tarea trasada
    dataset = datasetr.drop('trip_id',axis=1)
    dataset = dataset.drop('trip_headsign',axis=1)
    
    # Añadiedo la columna de distancia calculada a la siguiente parada
    
    direction_0, direction_1 = [],[]
    for ind,value in enumerate(dataset.iloc[:,10]):
        if dataset.iloc[ind,5] == 0:
            if dataset.iloc[ind,8] == 56:
                direction_0.append((value-dataset.iloc[ind,6])/1)
            else:
                direction_0.append((value-dataset.iloc[ind,6])/(56-dataset.iloc[ind,8]))
        else:
            if dataset.iloc[ind,8] == 57:
                direction_1.append((value-dataset.iloc[ind,6])/1)
            else:
                direction_1.append((value-dataset.iloc[ind,6])/(57-dataset.iloc[ind,8]))
    distance_next_stop_calculated = direction_0 + direction_1
    dataset['Distance_next_stop_calculated'] = np.array(distance_next_stop_calculated)
    
    # Descomponer el timestamp 
    dataset['year'] = dataset['timestamp'].dt.year
    dataset['month'] = dataset['timestamp'].dt.month
    dataset['day'] = dataset['timestamp'].dt.day
    dataset['hour'] = dataset['timestamp'].dt.hour
    dataset['second'] = dataset['timestamp'].dt.second
    dataset['minute'] = dataset['timestamp'].dt.minute
    
    dataset = dataset.drop('timestamp',axis=1)
    
    # Para cambiar el IN_PROGRESS 
    """for k,h in enumerate(dataset.iloc[:,3]):
    if h == 'IN_PROGRESS':
        dataset.iloc[k,3] = 1
    else:
        dataset.iloc[k,3] = 0
    """
    
    return dataset


In [8]:
data_ = data.sort_values(['vehicle_id','timestamp'])
datas = treatment_of_dataset(gen_dates(data.timestamp),bus_ids,data,data_[data_.direction_id == 0],data_[data_.direction_id == 1])

In [140]:
datas_copy = datas.fillna(value=0)

In [101]:
# Mierdero que tiene ahi armado

def trueque (datas):
    
    X = np.array(datas)
    #cambiar las phase's por ceros y uno
    # por el 4 ponga el número de columna donde está lo de phase
    for i in range(X.shape[0]):
        if X[i,3] == "IN_PROGRESS":
            X[i,3] =1 
        else:
            X[i,3] =0


    #Aqui lo mismo, por el 6 ponga el número de la columna de shape_dist_traveled
    restas_distancias =[]
    for i in range(X.shape[0]):
        if i==0:
            restas_distancias.append(int(X[i,5]))
        else:
            #Loca aca por year ponga el número de esa columna, lo mismo con minute
            if np.sum(abs(X[i, 11:16] - X[i-1, 11:16])) <=50 :
                restas_distancias.append(int(X[i,5]-X[i-1,5]))
            else: 
                restas_distancias.append(int(X[i,5]))
    X= np.concatenate((X,np.array([restas_distancias]).T),axis=1)
    
    return X

In [32]:
datas.head()

vehicle_id   latitude  longitude        phase  direction_id  \
0        7560  40.635662 -74.020229  IN_PROGRESS             0   
1        7560  40.635961 -74.019829  IN_PROGRESS             0   
2        7560  40.637289 -74.018488  IN_PROGRESS             0   
3        7560  40.638308 -74.017440  IN_PROGRESS             0   
4        7560  40.639967 -74.015637  IN_PROGRESS             0   

   shape_dist_traveled  stop_id  stop_sequence  dist_from_stop  \
0          3130.807395   305350             15       94.167148   
1          3172.240940   305350             15       52.733603   
2          3356.173073   308325             16      107.344252   
3          3500.303251   305353             17      208.453583   
4          3738.351868   308326             18      207.165236   

   average_distance  Distance_next_stop_calculated  year  month  day  hour  \
0      11768.693526                     210.680150  2011      4    4     4   
1      11768.693526                     209.669575  2011      4    4     4   
2      11768.693526                     210.313011  2011      4    4     4   
3      11768.693526                     212.010007  2011      4    4     4   
4      11768.693526                     211.324780  2011      4    4     4   

   second  minute  
0      45      18  
1      15      19  
2      45      19  
3      16      20  
4      45      20

In [33]:
X[0]

array([7560, 40.635661999999996, -74.0202288, 1, 0, 3130.8073952074,
       305350, 15, 94.1671476276601, 11768.69352575713, 210.68014952560318,
       2011, 4, 4, 4, 45, 18], dtype=object)

In [141]:
X = trueque(datas_copy)

In [142]:
y = X[:,8]
X = np.delete(X,8,1)

In [143]:
X.shape[0] , y.shape

(840940, (840940,))

In [144]:
porcent = int(X.shape[0]*0.70)
x_train = X[:porcent,:]
y_train = y[:porcent]

x_test = X[porcent:,:]
y_test = y[porcent:]

x_train.shape ,y_train.shape, x_test.shape, y_test.shape

((588658, 17), (588658,), (252282, 17), (252282,))

In [146]:
from sklearn.linear_model import LinearRegression

lr = LinearRegression()
lr.fit(x_train, y_train)
print("score en train = ",lr.score(x_train, y_train))
print("score en test = ", lr.score(x_test, y_test))

score en train =  0.387866351758
score en test =  0.545478148427


In [149]:
from sklearn.ensemble import RandomForestRegressor

rf = RandomForestRegressor()
rf.fit(x_train ,y_train)

print("score en train = ",rf.score(x_train, y_train))
print("score en test = ", rf.score(x_test, y_test))

score en train =  0.999527930709
score en test =  0.99531192397
